include/micm/util/vector_matrix.hpp

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2023 National Center for Atmospheric Research,
+//
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include <cassert>
+#include <cmath>
+#include <vector>
+
+namespace micm
+{
+
+  /// @brief A 2D array class with contiguous memory structured to encourage vectorization
+  ///
+  /// The memory layout groups rows into blocks whose size can be set such that for a single
+  /// column, the block of rows can fit in the vector register.
+  ///
+  /// The template arguments are the type of the matrix elements and the size of the number
+  /// of rows per block.
+  template<class T, std::size_t L>
+  class VectorMatrix
+  {
+    std::vector<T> data_;
+    std::size_t x_dim_;
+    std::size_t y_dim_;
+
+    friend class Proxy;
+    friend class ConstProxy;
+
+    class Proxy
+    {
+      VectorMatrix &matrix_;
+      std::size_t block_index_;
+      std::size_t row_index_;
+      std::size_t y_dim_;
+
+     public:
+      Proxy(VectorMatrix &matrix, std::size_t block_index, std::size_t row_index, std::size_t y_dim)
+          : matrix_(matrix),
+            block_index_(block_index),
+            row_index_(row_index),
+            y_dim_(y_dim)
+      {
+      }
+      Proxy &operator=(const std::vector<T> other)
+      {
+        assert(other.size() >= y_dim_ && "Matrix row size mismatch in assignment from vector");
```

Comment:
You'll need to do something like what I did in #84 to make these tests pass when we run them in release mode

https://github.com/NCAR/micm/pull/84/files#diff-78acddefff72c23089ee070cf969314ef0b55616b44fb5b03439896337a73913R40-R45

Reply:
done!